Source/WebCore/platform/graphics/CachedSubimage.h

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (C) 2023 Apple Inc.  All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "FloatRect.h"
+#include "ImageTypes.h"
+#include <wtf/FastMalloc.h>
+
+namespace WebCore {
+
+class GraphicsContext;
+class Image;
+class ImageBuffer;
+struct ImagePaintingOptions;
+
+class CachedSubimage {
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    static std::unique_ptr<CachedSubimage> create(Image&, GraphicsContext&, const FloatRect& destinationRect, const FloatRect& sourceRect, const ImagePaintingOptions&);
+
+    CachedSubimage(Ref<ImageBuffer>&&, const FloatSize& scaleFactor, const FloatRect& destinationRect, const FloatRect& sourceRect);
+
+    bool canBeUsed(GraphicsContext&, const FloatRect& destinationRect, const FloatRect& sourceRect) const;
+    void draw(GraphicsContext&, const FloatRect& destinationRect, const FloatRect& sourceRect);
+
+    // The tile size is usually 512 x 512, account for the 2x display resolution
+    // and make room for one more tile in every direction.
+    static constexpr float maxSide = 512 * 2 * 3;
```

Comment:
Yes I think we should. 

But we have a problem getting the `deviceScaleFactor` from the `GraphicsContext`. `GraphicsContext::scaleFactor()` returns the user transform multiplied by the `deviceScaleFactor`. Also `GraphicsContextCG::getCTM()` takes the argument `IncludeDeviceScale` whose values are confusing 
```
 enum IncludeDeviceScale { DefinitelyIncludeDeviceScale, PossiblyIncludeDeviceScale }
```
But even with passing PossiblyIncludeDeviceScale on WK1, I am still getting the user transform multiplied by the `deviceScaleFactor`. 

And this is why GraphicsContextCG uses the local inline function `getUserToBaseCTM()` to get the user CTM without the `deviceScaleFactor`. Also ControlMac::drawCell() has a duplicate of the same calculation to know whether to use an ImageBuffer or not.

Because this needs cleaning up the `GraphicsContext::scaleFactor()` and `GraphicsContext::getCTM()` and removing the duplicate code in GraphicsContextCG.cpp and ControlMac::drawCell(), I am going to address this in a separate patch.